core/os/os2/process_linux.odin

```diff
@@ -71,25 +307,303 @@ _Sys_Process_Attributes :: struct {}
+		found: bool
+		for dir in path_dirs {
+			executable_path = fmt.caprintf("%s/%s", dir, executable_name, temp_allocator())
+			fail: bool
+			if fail, errno = linux.faccessat(dir_fd, executable_path, linux.F_OK); errno == .NONE && !fail {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// check in cwd to match windows behavior
+			executable_path = fmt.caprintf("./%s", name, temp_allocator())
+			fail: bool
+			if fail, errno = linux.faccessat(dir_fd, executable_path, linux.F_OK); errno != .NONE || fail {
+				return process, .Not_Exist
+			}
+		}
+	} else {
+		executable_path = temp_cstring(executable_name) or_return
+	}
+
+	not_exec: bool
+	if not_exec, errno = linux.faccessat(dir_fd, executable_path, linux.F_OK | linux.X_OK); errno != .NONE || not_exec {
+		return process, errno == .NONE ? .Permission_Denied : _get_platform_error(errno)
+	}
+
+	// args and environment need to be a list of cstrings
+	// that are terminated by a nil pointer.
+	cargs := make([]cstring, len(desc.command) + 1, temp_allocator())
+	for command, i in desc.command {
+		cargs[i] = temp_cstring(command) or_return
+	}
+
+	// Use current process' environment if description didn't provide it.
+	env: [^]cstring
+	if desc.env == nil {
+		// take this process's current environment
+		env = raw_data(export_cstring_environment(temp_allocator()))
+	} else {
+		cenv := make([]cstring, len(desc.env) + 1, temp_allocator())
+		for env, i in desc.env {
+			cenv[i] = temp_cstring(env) or_return
+		}
+		env = &cenv[0]
+	}
+
+	// TODO: This is the traditional textbook implementation with fork.
+	//       A more efficient implementation with vfork:
+	//
+	//       1. retrieve signal handlers
+	//       2. block all signals
+	//       3. allocate some stack space
+	//       4. vfork (waits for child exit or execve); In child:
+	//           a. set child signal handlers
+	//           b. set up any necessary pipes
+	//           c. execve
+	//       5. restore signal handlers
+	//
+	pid: linux.Pid
+	if pid, errno = linux.fork(); errno != .NONE {
+		return process, _get_platform_error(errno)
+	}
+
+	STDIN  :: linux.Fd(0)
+	STDOUT :: linux.Fd(1)
+	STDERR :: linux.Fd(2)
+
+	if pid == 0 {
+		// in child process now
+		if desc.stdin != nil {
+			fd := linux.Fd(fd(desc.stdin))
+			if _, errno = linux.dup2(fd, STDIN); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+		if desc.stdout != nil {
+			fd := linux.Fd(fd(desc.stdout))
+			if _, errno = linux.dup2(fd, STDOUT); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+		if desc.stderr != nil {
+			fd := linux.Fd(fd(desc.stderr))
+			if _, errno = linux.dup2(fd, STDERR); errno != .NONE {
+				linux.exit(1)
+			}
+		}
+
+		if errno = linux.execveat(dir_fd, executable_path, &cargs[0], env); errno != .NONE {
+			print_error(stderr, _get_platform_error(errno), string(executable_path))
+			panic("execve failed to replace process")
+		}
+		unreachable()
+	}
+
+	process.pid = int(pid)
+	process.handle = PIDFD_UNASSIGNED
+	return
```

Comment:
I changed this to `exit`, but the more I look at it the more the logic is wrong.   It should be aborting with something like an illegal instruction.  This gives us 2 options:

 - The kernel will get the message out for us no matter what we did to the pipes.  It probably just writes directly to `/dev/stderr` or something.  It is up to the user to handle the state they get back from `process_wait` and `process_start` still returns success.
 - Catch a signal from the child sent via `kill(get_ppid(), ...)`

For the signal path, either we lock and take over a signal temporarily, or we have to take full ownership of an arbitrary signal number that the child can report to.  Both have their complications, so maybe best to just leave it to the user.

What's the best way to hard fault here?  I cannot find an `abort` equivalent that doesn't try to print something.

Reply:
`intrinsics.ud2`. I think it's called something else though, but basically on x86 it should produce a UD2 instruction which will be interpreted in a debugger as an invalid instruction, pretty telling for something that goes "expectedly wrong". It might be `intrinsics.trap()`, or `debug_trap()`, but honestly I don't remember, it's one of those low-level procedures.

As for the error handling, I do have a couple of ideas, but none of them are good.

I think one of the ways to resolve it is to use user-defined signals `SIG_USR+n`, for custom error codes. The child would send one of these to its parent and abort. Then it would send a cross-process message saying that it's about to be started. The parent would wait on that message after calling `fork`, and after the wait it would examine the status of the child process and return the according status. We can guarantee to be able to do this upto the point of `execve`, but we won't be able to handle `execve` errors (because we cant synchronize the message in the case of success, at that point the code is out of our control).

The issues with this approach are:

1. It's intrusive. We're taking up user signals for error handling purposes
2. We can't handle errors returned by execve

The first issue is probably not an issue, because `os2` doesn't provide signals to users, but it can be a problem in an application that uses both `os2` and system-specific functions. Nevertheless, it can be resolved by using literally any other IPC mechanism. I'm not sure how to communicate a child process an inherited pipe, but I think communicating error codes via an extra pipe that will be closed right before execve is not a bad choice.

The second issue is mainly an issue because normally execve would return an error that says that a file can't be found. This is trivially fixed by using `fexecve` and opening file descriptor as a separate operation.

`fexecve` could still return us an error, and here we give up and just debug trap our process. I think that's the best error handling that we can afford.

I'd need to look into how other similar programming languages handle these errors. Particularly curious about zig, because it's also a type of language that uses raw syscalls in their standard library.

Reply:
That's simple enough.  I like the idea of a temporary pipe instead of playing with signals.   `fexecve` is libc.  I'm curious how they implement that..

Reply:
WHAT!?  `execveat` can act like an `fexecve` if the pathname is blank an is provided `AT_EMPTY_PATH`.  I didin't know that...

Reply:
Here's how zig does it: https://github.com/ziglang/zig/blob/master/lib/std/process/Child.zig#L646

My intuition about setting up a separate pipe was correct

Reply:
That's interesting.  They hold onto it and read it during the wait routine.  In my head, I thought pipe on the stack, child sends a single byte status, and parent spins waiting for it before leaving `process_start`.

Reply:
There are still these funny little issues in sys/linux that are going to take a few months to fix all of them. That's kind of how manual binding works :)